### PR TITLE
Give up on escalating commands early when standing up

### PR DESCRIPTION
**Pull request by tylerkaraszewski**
cc @rafecolton 

### Details
In this PR: https://github.com/Expensify/Bedrock/pull/1241 we noticed a bit of a hole. As one leader shuts down and another takes over, any other node bedsides those two will continually look for a new leader, and when it finds it, continue escalating commands to the new leader.

However, on the node that is becoming leader, `runOnLeader` will never find that new leader, because it *is* leader. It will then run out the entire 5 second timeout before these commands that it was attempting to run get requeued to run locally.

This adds a quick check to see if, when there's no leader address, *we're the leader* and if so, returns without waiting the full 5 seconds so these commands can run sooner.

### Fixed Issues
https://github.com/Expensify/Expensify/issues/193926

### Tests
Tested in dev (clustertest gracefulfailovertest) here are logs where this happens:
```
2022-04-22T23:21:42.498677+00:00 expensidev2004 bedrock10004: OCtwIv (BedrockCore.cpp:81) peekCommand [worker7] [dbug] Peeking at 'idcollision r_11579_r' with priority: 500
2022-04-22T23:21:42.498694+00:00 expensidev2004 bedrock10004: OCtwIv (SQLite.cpp:316) beginTransaction [worker7] [dbug] [concurrent] Beginning transaction
2022-04-22T23:21:42.498946+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:2527) SQuery [worker7] [dbug] BEGIN CONCURRENT
2022-04-22T23:21:42.498970+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:2527) SQuery [worker7] [dbug] PRAGMA query_only = true;
2022-04-22T23:21:42.505504+00:00 expensidev2004 bedrock10004: OCtwIv (BedrockCore.cpp:98) peekCommand [worker7] [dbug] Plugin 'TestPlugin' peeked command 'idcollision r_11579_r'
2022-04-22T23:21:42.505527+00:00 expensidev2004 bedrock10004: OCtwIv (BedrockCore.cpp:101) peekCommand [worker7] [dbug] Command 'idcollision r_11579_r' not finished in peek, re-queuing.
2022-04-22T23:21:42.505539+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:2527) SQuery [worker7] [dbug] PRAGMA query_only = false;
2022-04-22T23:21:42.505551+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:2527) SQuery [worker7] [dbug] ROLLBACK
2022-04-22T23:21:42.505712+00:00 expensidev2004 bedrock10004: OCtwIv (SQLite.cpp:748) rollback [worker7] [dbug] Transaction rollback with 0 queries attempted, 0 served from cache.
2022-04-22T23:21:42.505871+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:1729) S_socket [worker7] [info] DNS lookup took 0ms for '0.0.0.0'.
2022-04-22T23:21:42.505902+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:1741) S_socket [worker7] [info] Resolved 0.0.0.0 to ip: 0.0.0.0.
2022-04-22T23:21:42.506241+00:00 expensidev2004 bedrock10004: OCtwIv (SQLiteClusterMessenger.cpp:52) waitForReady [worker7] [info] [HTTPESC] Socket disconnected while waiting to be ready (send).
2022-04-22T23:21:43.007002+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:1729) S_socket [worker7] [info] DNS lookup took 0ms for '0.0.0.0'.
2022-04-22T23:21:43.007048+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:1741) S_socket [worker7] [info] Resolved 0.0.0.0 to ip: 0.0.0.0.
2022-04-22T23:21:43.007062+00:00 expensidev2004 bedrock10004: OCtwIv (SQLiteClusterMessenger.cpp:52) waitForReady [worker7] [info] [HTTPESC] Socket disconnected while waiting to be ready (send).
2022-04-22T23:21:43.507961+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:1729) S_socket [worker7] [info] DNS lookup took 0ms for '0.0.0.0'.
2022-04-22T23:21:43.508055+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:1741) S_socket [worker7] [info] Resolved 0.0.0.0 to ip: 0.0.0.0.
2022-04-22T23:21:43.508097+00:00 expensidev2004 bedrock10004: OCtwIv (SQLiteClusterMessenger.cpp:52) waitForReady [worker7] [info] [HTTPESC] Socket disconnected while waiting to be ready (send).
2022-04-22T23:21:44.009289+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:1729) S_socket [worker7] [info] DNS lookup took 0ms for '0.0.0.0'.
2022-04-22T23:21:44.009342+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:1741) S_socket [worker7] [info] Resolved 0.0.0.0 to ip: 0.0.0.0.
2022-04-22T23:21:44.009384+00:00 expensidev2004 bedrock10004: OCtwIv (SQLiteClusterMessenger.cpp:52) waitForReady [worker7] [info] [HTTPESC] Socket disconnected while waiting to be ready (send).
2022-04-22T23:21:44.510618+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:1729) S_socket [worker7] [info] DNS lookup took 0ms for '0.0.0.0'.
2022-04-22T23:21:44.510678+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:1741) S_socket [worker7] [info] Resolved 0.0.0.0 to ip: 0.0.0.0.
2022-04-22T23:21:44.510704+00:00 expensidev2004 bedrock10004: OCtwIv (SQLiteClusterMessenger.cpp:52) waitForReady [worker7] [info] [HTTPESC] Socket disconnected while waiting to be ready (send).
2022-04-22T23:21:45.013013+00:00 expensidev2004 bedrock10004: OCtwIv (SQLiteClusterMessenger.cpp:88) runOnLeader [worker7] [info] [HTTPESC] I'm the leader now! Exiting early.
2022-04-22T23:21:45.013082+00:00 expensidev2004 bedrock10004: OCtwIv (BedrockServer.cpp:1045) worker [worker7] [info] Couldn't escalate command idcollision r_11579_r to leader. We are in state: FOLLOWING
2022-04-22T23:21:45.013111+00:00 expensidev2004 bedrock10004: OCtwIv (BedrockServer.cpp:770) worker [worker7] [info] Dequeued command idcollision r_11579_r (cluster_node_1#103) in worker, 21 commands in  queue.
2022-04-22T23:21:45.724618+00:00 expensidev2004 bedrock10004: OCtwIv (BedrockCore.cpp:81) peekCommand [worker7] [dbug] Peeking at 'idcollision r_11579_r' with priority: 500
2022-04-22T23:21:45.724930+00:00 expensidev2004 bedrock10004: OCtwIv (SQLite.cpp:316) beginTransaction [worker7] [dbug] [concurrent] Beginning transaction
2022-04-22T23:21:45.724945+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:2527) SQuery [worker7] [dbug] BEGIN CONCURRENT
2022-04-22T23:21:45.724960+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:2527) SQuery [worker7] [dbug] PRAGMA query_only = true;
2022-04-22T23:21:45.731253+00:00 expensidev2004 bedrock10004: OCtwIv (BedrockCore.cpp:98) peekCommand [worker7] [dbug] Plugin 'TestPlugin' peeked command 'idcollision r_11579_r'
2022-04-22T23:21:45.731277+00:00 expensidev2004 bedrock10004: OCtwIv (BedrockCore.cpp:101) peekCommand [worker7] [dbug] Command 'idcollision r_11579_r' not finished in peek, re-queuing.
2022-04-22T23:21:45.731290+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:2527) SQuery [worker7] [dbug] PRAGMA query_only = false;
2022-04-22T23:21:45.731301+00:00 expensidev2004 bedrock10004: OCtwIv (BedrockCore.cpp:183) processCommand [worker7] [dbug] Processing 'idcollision r_11579_r'
2022-04-22T23:21:45.741957+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:2527) SQuery [worker7] [dbug] SELECT MAX(id) FROM test
2022-04-22T23:21:45.742035+00:00 expensidev2004 bedrock10004: OCtwIv (SQLite.cpp:261) _sqliteLogCallback [worker7] [info] {SQLITE} Code: 1, Message: no such table: test in "SELECT MAX(id) FROM test"
2022-04-22T23:21:45.743215+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:2527) SQuery [worker7] [dbug] PRAGMA schema_version;
2022-04-22T23:21:45.743250+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:2527) SQuery [worker7] [dbug] INSERT INTO TEST VALUES(148, '');
2022-04-22T23:21:45.743266+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:2527) SQuery [worker7] [dbug] PRAGMA schema_version;
2022-04-22T23:21:45.743278+00:00 expensidev2004 bedrock10004: OCtwIv (BedrockCore.cpp:209) processCommand [worker7] [dbug] Plugin 'TestPlugin' processed command 'idcollision r_11579_r'
2022-04-22T23:21:45.743289+00:00 expensidev2004 bedrock10004: OCtwIv (BedrockCore.cpp:231) processCommand [worker7] [info] Processed '756' for 'idcollision r_11579_r'.
2022-04-22T23:21:45.746102+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:2527) SQuery [worker7] [dbug] INSERT INTO journal0001 VALUES (150, 'INSERT INTO TEST VALUES(148, '''');', 'E2166C3F79CF13DB09AFBA88F70669CB39AC152C' )
2022-04-22T23:21:45.746206+00:00 expensidev2004 bedrock10004: OCtwIv (SQLite.cpp:583) prepare [worker7] [dbug] Prepared transaction
2022-04-22T23:21:45.746223+00:00 expensidev2004 bedrock10004: OCtwIv (SQLite.cpp:622) commit [worker7] [dbug] Committing transaction
2022-04-22T23:21:45.746232+00:00 expensidev2004 bedrock10004: OCtwIv (libstuff.cpp:2527) SQuery [worker7] [dbug] COMMIT
```

You can see that this command is dequeued, attempted to escalate to leader, but then we notice we are leader, and the command is run locally, and the whole thing happens in about 3 seconds (saving some user 2 seconds of waiting)

### CODE REVIEW

**Review by rafecolton — APPROVED** [review]
Nice, simple fix! 🙌🏼

File: sqlitecluster/SQLiteClusterMessenger.cpp [hunk 1]
```diff
@@ -82,6 +82,14 @@ bool SQLiteClusterMessenger::runOnLeader(BedrockCommand& command) {
     while (chrono::steady_clock::now() < (start + 5s) && !sent) {
         string leaderAddress = _node->leaderCommandAddress();
         if (leaderAddress.empty()) {
+            // If there's no leader, it's possible we're supposed to be the leader. In this case, we can exit early.
+            auto myState = _node->getState();
+            if (myState == STCPNode::LEADING || myState == STCPNode::STANDINGUP) {
+                SINFO("[HTTPESC] I'm the leader now! Exiting early.");
```
Comment (bondydaa) [inline sqlitecluster/SQLiteClusterMessenger.cpp:88]: 🤣

**Review by bondydaa — APPROVED** [review]
👍 looks good though I'm wondering based on these log lines
```
2022-04-22T23:21:45.013013+00:00 expensidev2004 bedrock10004: OCtwIv (SQLiteClusterMessenger.cpp:88) runOnLeader [worker7] [info] [HTTPESC] I'm the leader now! Exiting early.
2022-04-22T23:21:45.013082+00:00 expensidev2004 bedrock10004: OCtwIv (BedrockServer.cpp:1045) worker [worker7] [info] Couldn't escalate command idcollision r_11579_r to leader. We are in state: FOLLOWING
```

Should we maybe "re-set" the `state` variable within the `worker` method loop since based on those log lines it appears initially the worker thread "thinks" it's a `FOLLOWER`, I'm guessing because we initially set the state here https://github.com/Expensify/Bedrock/blob/1ea2cf66c60f8771e16545f15a78cb03c2e0c4d4/BedrockServer.cpp#L848?

Or I guess that'd be the alternative change to what you've done here? I guess this is a bit easier to do than trying to update all the logic around `state` in the worker method.

File: sqlitecluster/SQLiteClusterMessenger.cpp (hunk 1 above)
Comment (bondydaa) [reply sqlitecluster/SQLiteClusterMessenger.cpp:88]: makes me think of "i'm a big kid now!" commercial https://youtu.be/ztTHLDhmB58?t=25

Comment (tylerkaraszewski) [reply sqlitecluster/SQLiteClusterMessenger.cpp:88]: Hah, I was going more for this: https://www.youtube.com/watch?v=dvA-mimf2yg

**Pull request merged**